File: src/test/java/edu/colorado/plv/fixr/TestSlicing.java
package edu.colorado.plv.fixr;

import static org.junit.Assert.assertTrue;

import java.util.Collection;
import java.util.Collections;

import edu.colorado.plv.fixr.slicing.APISlicer;
import edu.colorado.plv.fixr.slicing.MethodPackageSeed;
import soot.Body;
import soot.SootClass;

public abstract class TestSlicing extends TestClassBase {

	public abstract Collection<String> getPackages();
	
	public void testSlice(String methodName) 
	{		
		SootClass testClass = this.getTestClass();
		SootClass resClass = this.getResClass();
		
		assertTrue(testClass != null);
		assertTrue(resClass!= null);
		
		Body bodyToSlice = testClass.getMethodByName(methodName).retrieveActiveBody();
		Body expectedRes = resClass.getMethodByName(APISlicer.getSlicedMethodName(methodName)).retrieveActiveBody();

		APISlicer slicer = new APISlicer(bodyToSlice);
		Body slicedBody = slicer.slice(new MethodPackageSeed(getPackages()));
		
//		// DEBUG
//		{
//			UnitGraph u1 = new BriefUnitGraph(bodyToSlice);
//			UnitGraph u2 = new BriefUnitGraph(slicedBody);
//			SootHelper.dumpToDot(u1, bodyToSlice, "/tmp/toslice.dot");
//			SootHelper.dumpToDot(u2, slicedBody, "/tmp/sliced.dot");			
//		}
		
		assertTrue(SootHelper.compareBodies(expectedRes, slicedBody));		
	}

}

File: src/test/scala/edu/colorado/plv/fixr/tests/slicing/TestSlicing.scala
package edu.colorado.plv.fixr.tests.slicing

import edu.colorado.plv.fixr.tests.TestClassBase
import edu.colorado.plv.fixr.slicing.APISlicer
import soot.Body
import edu.colorado.plv.fixr.slicing.MethodPackageSeed
import scala.collection.JavaConversions._
import org.scalatest._
import edu.colorado.plv.fixr.SootHelper
import edu.colorado.plv.fixr.SootHelper
import edu.colorado.plv.fixr.slicing.APISlicer
import edu.colorado.plv.fixr.tests.TestClassBase
import edu.colorado.plv.fixr.SootHelper
import edu.colorado.plv.fixr.slicing.APISlicer
import edu.colorado.plv.fixr.tests.TestClassBase
import edu.colorado.plv.fixr.SootHelper
import edu.colorado.plv.fixr.slicing.APISlicer
import edu.colorado.plv.fixr.tests.TestClassBase
import edu.colorado.plv.fixr.SootHelper
import edu.colorado.plv.fixr.slicing.APISlicer
import edu.colorado.plv.fixr.tests.TestClassBase

/**
  * Test the slicing of a method.
  * @author: Sergio Mover
  * @mail: sergio.mover@colorado.edu
  */
abstract class TestSlicing(classPath : String, testClassName : String,
  resClassName : String)
    extends TestClassBase(classPath, testClassName, resClassName) {

  /**
    * Get a list of packages to be included in the slice
    */
  def getPackages() : List[String];

  /**
    * Test the slice of a method
    */
  def testSlice(methodName : String) : Unit = {
    assert(testClass != null)
    assert(resClass != null)

    val bodyToSlice : Body = testClass.getMethodByName(methodName).retrieveActiveBody();
    val expectedRes : Body = resClass.getMethodByName(APISlicer.getSlicedMethodName(methodName)).retrieveActiveBody();

    val slicer : APISlicer = new APISlicer(bodyToSlice);
    val slicedBody : Body = slicer.slice(new MethodPackageSeed(getPackages()));

    assert(null != slicedBody) /* at least one seed in the test*/
    /* obtained and expected results must be the same */
    assert(SootHelper.compareBodies(expectedRes, slicedBody));
  }
}
